lib/OpenQA/WebAPI/Plugin/RateLimits.pm

```diff
@@ -0,0 +1,43 @@
+# Copyright (C) 2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::WebAPI::Plugin::RateLimits;
+use Mojo::Base 'Mojolicious::Plugin', -signatures;
+
+
+use Time::Seconds;
+
+sub register {
+    my ($self, $app) = @_;
+
+    $app->hook(
+        around_action => sub ($next, $c, $action, $last) {
+            return $next->() unless $last;
+            # Allow n queries per minute, per user (if logged in)
+            my $route    = $c->current_route;
```

Comment:
Below this like i'd just put an if that short circuits all the other rate limit checks.
```perl
if ($route && (my $limit = $c->app->config->{rate_limits}{$route})) {
  ...
```